http/src/main/scala/org/http4s/blaze/http/http2/Http2Settings.scala

```diff
@@ -0,0 +1,224 @@
+    val MAX_FRAME_SIZE = 16384                     // octets (2^14)
+    val MAX_HEADER_LIST_SIZE = Integer.MAX_VALUE   // octets (infinite)
+  }
+
+  final case class Setting(code: Int, value: SettingValue) {
+
+    /** Get the SettingKey associated with this setting */
+    def key: SettingKey = settingKey(code)
+
+    /** Get the human readable name of this setting */
+    def name: String = key.name
+
+    override def toString = s"$name(0x${Integer.toHexString(code)}, $value"
+  }
+
+  /** Create a new [[Http2Settings]] initialized with the protocol defaults */
+  val default: ImmutableHttp2Settings = ImmutableHttp2Settings(
+    headerTableSize = DefaultSettings.HEADER_TABLE_SIZE,
+    initialWindowSize = DefaultSettings.INITIAL_WINDOW_SIZE,
+    pushEnabled = DefaultSettings.ENABLE_PUSH, // initially enabled
+    maxConcurrentStreams = DefaultSettings.MAX_CONCURRENT_STREAMS, // initially unbounded
+    maxFrameSize = DefaultSettings.MAX_FRAME_SIZE,
+    maxHeaderListSize = DefaultSettings.MAX_HEADER_LIST_SIZE
+  )
+
+  private def settingKey(id: Int): SettingKey =
+    settingsMap.getOrElse(id, SettingKey(id, s"UNKNOWN_SETTING(0x${Integer.toHexString(id)})"))
+
+  private val settingsMap = new mutable.HashMap[Int, SettingKey]()
+
+  private def makeKey(code: Int, name: String): SettingKey = {
+    val k = SettingKey(code, name)
+    settingsMap += code -> k
+    k
+  }
+
+  /** Helper for extracting invalid settings
+    *
+    * @see https://tools.ietf.org/html/rfc7540#section-6.5.2
+    */
+  object InvalidSetting {
+    def unapply(setting: Setting): Option[Http2Exception] = setting match {
+      case MAX_FRAME_SIZE(size) if 16384 > size || size > 16777215 =>
+        Some(Http2Exception.PROTOCOL_ERROR.goaway(s"Invalid MAX_FRAME_SIZE: $size"))
+
+      case Setting(code, value) if value < 0 =>
+        val ex = Http2Exception.PROTOCOL_ERROR.goaway(s"Integer overflow for setting ${setting.name}: ${value}")
+        Some(ex)
+
+      case _ => None
+    }
+  }
+
+  final case class SettingKey(code: Int, name: String) {
+    override def toString = name
+
+    /** Create a new `Setting` with the provided value */
+    def apply(value: SettingValue): Setting = Setting(code, value)
+
+    /** Extract the value from the key */
+    def unapply(setting: Setting): Option[SettingValue] =
+      if (setting.code == code) Some(setting.value)
+      else None
+  }
+
+  val HEADER_TABLE_SIZE      = makeKey(0x1, "SETTINGS_HEADER_TABLE_SIZE")
+  val ENABLE_PUSH            = makeKey(0x2, "SETTINGS_ENABLE_PUSH")
+  val MAX_CONCURRENT_STREAMS = makeKey(0x3, "SETTINGS_MAX_CONCURRENT_STREAMS")
+  val INITIAL_WINDOW_SIZE    = makeKey(0x4, "SETTINGS_INITIAL_WINDOW_SIZE")
+  val MAX_FRAME_SIZE         = makeKey(0x5, "SETTINGS_MAX_FRAME_SIZE")
+  val MAX_HEADER_LIST_SIZE   = makeKey(0x6, "SETTINGS_MAX_HEADER_LIST_SIZE")
+}
+
+/** Internal mutable representation of the [[Http2Settings]] */
+private final case class MutableHttp2Settings private(
```

Comment:
What are we getting out of being a case class if it's mutable?

Reply:
Probably initially a good `toString`, but nothing that I can think of now. I'll remove it.